### PR TITLE
orderbook: export NewDepth(), add methods to orderbook.Unsafe type

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Exports NewDepth functionality for testing and implementation outside of the library. 

Adds methods for basic checks and data fetching directly from the linked list. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/orderbook/unsafe.go [hunk 1]
```diff
@@ -62,3 +66,100 @@ func (d *Depth) GetUnsafe() Unsafe {
...
+// GetBestBid returns the top bid price
+func (src *Unsafe) GetBestBid() (float64, error) {
+	bid, err := src.getBidLiquidity()
+	if err != nil {
+		return 0, fmt.Errorf("get orderbook best bid price %w", err)
+	}
+	return bid.Value.Price, nil
+}
+
+// GetBestBid returns the top bid price
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:91]:
```suggestion
// GetBestAsk returns the top ask price
```

File: exchanges/orderbook/unsafe.go [hunk 2]
```diff
@@ -62,3 +66,100 @@ func (d *Depth) GetUnsafe() Unsafe {
...
+	}
+	ask, err = src.getAskLiquidity()
+	if err != nil {
+		return nil, nil, err
+	}
+	return ask, bid, nil
+}
+
+// GetBestBid returns the average between the top bid and top ask.
+func (src *Unsafe) GetMidPrice() (float64, error) {
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:128]: comment

File: exchanges/orderbook/unsafe.go [hunk 3]
```diff
@@ -62,3 +66,100 @@ func (d *Depth) GetUnsafe() Unsafe {
 		LastUpdated:    &d.options.lastUpdated,
 	}
 }
+
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:69]: Do any of the below functions need to utilise locks? I recognise that this is `Unsafe` and locking has been exported, but it's a potential issue. I note that you still use locks with `IsStreaming`

File: exchanges/orderbook/unsafe_test.go [hunk 4]
```diff
@@ -26,3 +29,158 @@ func TestUnsafe(t *testing.T) {
...
+	if !errors.Is(err, nil) {
+		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
+	}
+}
+
+func TestGetBestBid(t *testing.T) {
+	t.Parallel()
+	d := NewDepth(unsafeID)
+	unsafe := d.GetUnsafe()
+	_, err := unsafe.GetBestBid()
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe_test.go:69]: Looks like github dislikes your `err` use here and below

File: exchanges/orderbook/unsafe.go [hunk 5]
```diff
@@ -62,3 +66,100 @@ func (d *Depth) GetUnsafe() Unsafe {
...
+	}
+	top := bid.Value.Amount - ask.Value.Amount
+	bottom := bid.Value.Amount + ask.Value.Amount
+	return top / bottom, nil
+}
+
+// IsStreaming returns if the orderbook is updated by a streaming protocol and
+// is most likely more up to date than that of a REST protocol update.
+// NOTE: This locks down the book for inspection.
+func (src *Unsafe) IsStreaming() bool {
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:160]: Add test for this func

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Looks like some nice helper funcs ⭐ Just a couple of tiny nits

File: exchanges/orderbook/unsafe.go (hunk 3 above)
Comment (shazbert) [reply exchanges/orderbook/unsafe.go:69]: Yeah they do, but they are just helpers as well, any inspection on the linked list externally would require a lock. I will remove the lock for the IsStreaming function. It's up to the caller to provide locking protection. And I will export everthing.

File: exchanges/orderbook/unsafe.go [hunk 6]
```diff
@@ -90,43 +90,45 @@ func (src *Unsafe) GetBestBid() (float64, error) {
...
+func (src *Unsafe) GetAskLiquidity() (*Node, error) {
+	n := *src.AskHead
+	if n == nil {
 		return nil, fmt.Errorf("bid %w", errNoLiquidity)
 	}
-	return *src.AskHead, nil
+	return n, nil
 }
 
 // getLiquidity checks and returns nodes to the top bids and asks
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:116]: CoMmEnT

File: exchanges/orderbook/unsafe.go [hunk 7]
```diff
@@ -90,43 +90,45 @@ func (src *Unsafe) GetBestBid() (float64, error) {
...
+	if n == nil {
 		return nil, fmt.Errorf("bid %w", errNoLiquidity)
 	}
-	return *src.BidHead, nil
+	return n, nil
 }
 
-func (src *Unsafe) getAskLiquidity() (*Node, error) {
-	if *src.AskHead == nil {
+func (src *Unsafe) GetAskLiquidity() (*Node, error) {
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:108]: Comment

File: exchanges/orderbook/unsafe.go [hunk 8]
```diff
@@ -90,43 +90,45 @@ func (src *Unsafe) GetBestBid() (float64, error) {
...
+	ask, err := src.GetAskLiquidity()
 	if err != nil {
 		return 0, fmt.Errorf("get orderbook best bid price %w", err)
 	}
 	return ask.Value.Price, nil
 }
 
-func (src *Unsafe) getBidLiquidity() (*Node, error) {
-	if *src.BidHead == nil {
+func (src *Unsafe) GetBidLiquidity() (*Node, error) {
```
Comment (gloriousCode) [inline exchanges/orderbook/unsafe.go:100]: Comment

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes. Found some _critical_ issues though

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those comments

File: exchanges/orderbook/unsafe.go [hunk 9]
```diff
@@ -62,3 +66,103 @@ func (d *Depth) GetUnsafe() Unsafe {
...
+		return 0, fmt.Errorf("get orderbook best bid price %w", err)
+	}
+	return bid.Value.Price, nil
+}
+
+// GetBestAsk returns the top ask price
+func (src *Unsafe) GetBestAsk() (float64, error) {
+	ask, err := src.GetAskLiquidity()
+	if err != nil {
+		return 0, fmt.Errorf("get orderbook best bid price %w", err)
```
Comment (thrasher-) [inline exchanges/orderbook/unsafe.go:95]:
```suggestion
		return 0, fmt.Errorf("get orderbook best ask price %w", err)
```

File: exchanges/orderbook/unsafe.go [hunk 10]
```diff
@@ -62,3 +66,103 @@ func (d *Depth) GetUnsafe() Unsafe {
...
+		return nil, fmt.Errorf("bid %w", errNoLiquidity)
+	}
+	return n, nil
+}
+
+// GetAskLiquidity gets the head node for the ask liquidity
+func (src *Unsafe) GetAskLiquidity() (*Node, error) {
+	n := *src.AskHead
+	if n == nil {
+		return nil, fmt.Errorf("bid %w", errNoLiquidity)
```
Comment (thrasher-) [inline exchanges/orderbook/unsafe.go:113]:
```suggestion
		return nil, fmt.Errorf("ask %w", errNoLiquidity)
```

**Review by thrasher- — CHANGES_REQUESTED** [review]
Two CRITICAL nits 😆 Really nice methods!

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes!

**Pull request merged**